lib/ansible/modules/windows/win_service.ps1

```diff
@@ -371,6 +456,88 @@ Function Set-ServiceState($svc, $wmi_svc, $state) {
     }
 }
 
+Function Set-ServiceRecovery($name, $reset_interval, $actions) {
+
+    # https://stackoverflow.com/questions/36462623/what-reg-binary-to-set-for-failureaction-for-service
+
+    # dwResetPeriod 2C 01 00 00
+
+    # lpRebootMsg 00 00 00 00
+
+    # lpCommand 00 00 00 00
+
+    # cActions 03 00 00 00 (never changes)
+
+    # lpsaActions 14 00 00 00 (never changes)
+
+    # https://docs.microsoft.com/en-us/windows/desktop/api/winsvc/ns-winsvc-_sc_action
+
+    # 01 00 00 00 60 EA 00 00
+    # 01 00 00 00 60 EA 00 00
+    # 01 00 00 00 60 EA 00 00
+    # actions:
+    #   0 for nothing
+    #   1 for restart service
+    #   2 for computer reboot
+    #   3 for run command (not implemented)
+    #   reset_period: 300 (seconds)
+    #   first_failure:
+    #     action: 1
+    #     timer: 60000 (ms)
+    #   second_failure: 1
+    #     action: 1
+    #     timer: 60000 (ms)
+    #   subsequent_failures: 1
+    #     action: 1
+    #     timer: 60000 (ms)
+    # 2c,01,00,00,00,00,00,00,00,00,00,00,03,00,00,00,14,00,00,00,01,00,00,00,60,ea,00,00,01,00,00,00,60,ea,00,00,01,00,00,00,60,ea,00,00
+
+    # Default untouched actions
+    $lp_reboot_msg = '00000000'
+    $lp_command = '00000000'
+    $ca_actions = '03000000'
+    $lpsaActions = '14000000'
+
+    $reset_fail_count_after = ReversedHexValue $reset_interval
+
+    if ($actions.Count -eq 3) {
```

Comment:
Why have a hard count here, we should just fill the remaining options with Take No Action if it's less than 3. Maybe fail if it's more than 3. Also if it's an empty list we should probably set Take No Action for all 3.

Reply:
In https://github.com/ansible/ansible/pull/56445#discussion_r345020105 you mentioned that 3 is just a GUI limitation. And if the third element is for the subsequent failures, does that mean that the last option in the list (3rd in this case), the last action set, is meant for the subsequent actions for failures?
But it's still confusing though as your top comment would say to fill the actions up as 'Take no action', meaning, if you just set 1 (or even till 2), no action would be taken afterwards. But if you go to the 3rd action, it'll repeat for all the next. Does that mean if you set the action on entry 6 (in theory), the 6th time would count for 6 or more subsequent failures to take that same action?

Reply:
> does that mean that the last option in the list (3rd in this case), the last action set, is meant for the subsequent actions for failures?

That is correct, if you read the [Remarks](https://docs.microsoft.com/en-us/windows/win32/api/winsvc/ns-winsvc-service_failure_actionsw) for the failure actions structure you can see the following line

> If N is greater than cActions, the service controller repeats the last action in the array.

So say you have 2 actions define, 1 to run a command and the 2nd to restart the service it will do 1 then 2 and then subsequently repeat 2 forever.

> But it's still confusing though as your top comment would say to fill the actions up as 'Take no action

Sorry that's my misunderstanding there and I was wrong when I said that. Here is the various scenarios I see with `recovery_actions`:

* `recovery_actions: null` (or just not set): Do not do anything, keep the existing actions as is
* `recovery_actions: []`: Make sure we remove all recovery actions if any exist
* `recovery_actions: [action1, action2]`: Make sure we have 2 actions set, that make action1 and action2
* `recovery_actions: [action1, action2, action3]`: Same as above but we have 3 actions now
* `recovery_actions: [1, 2, 3, 4]`: For now we need to fail as `sc.exe` doesn't support this but in the future when we move to a PInvoke impl we can remove this limit

I've tested it out and `sc.exe failure` can be used to set 1, 2, or 3 actions not just 0 or 3. So the code just needs to make sure we check the existing settings properly and then set it if there is a change.

> Does that mean if you set the action on entry 6 (in theory), the 6th time would count for 6 or more subsequent failures to take that same action?

Yes that is correct, althrough `sc.exe failure` only allows you to set 3 you *can* set more by using the Win32 API which is a lot more involved.

Reply:
Just an FYI, I was playing around with it a bit more yesterday and created a barebones cmdlet to overcome the 3 action limit https://gist.github.com/jborean93/889288b56087a2c5def7fa49b6a8a0ad. Through this I was able to verify that I can set 4 or more actions and Windows will continue to work through them and finally repeat the last action.

Reply:
At this point, we might as well just drop this PR and create a separate module that implements the Win32 API for Windows services recovery actions then. 😃

Reply:
Up to you, I don't really want to take away from all the work you've done so far. If we were to make it into a separate module we can close this one. If we wanted to include the functionality into this module we can do 2 things

1. Keep moving forward with this and at a future point in time implement the PInvoke stuff in a separate PR
2. Just do the PInvoke stuff in 1 PR

Honestly I will leave it up to you, you've done a lot of work here and I don't take away from what's currently here.

Reply:
Ok, I'll think about it over the week-end!
In any case, I needed these changes internally in my company, so the work is not lost and we use a local version in our repo.